python/deltalake/schema.py

```diff
@@ -17,34 +17,91 @@
 
-    Returns:
-        A PyArrow Table and the inferred schema for the Delta Table
+    Args
+        schema: Source schema
+        large_dtypes: If True, the pyarrow schema is casted to large_dtypes
     """
+    dtype_map = {
+        pa.uint8(): pa.int8(),
+        pa.uint16(): pa.int16(),
+        pa.uint32(): pa.int32(),
+        pa.uint64(): pa.int64(),
+    }
+    if large_dtypes:
+        dtype_map = {
+            **dtype_map,
+            **{pa.string(): pa.large_string(), pa.binary(): pa.large_binary()},
+        }
+    else:
+        dtype_map = {
+            **dtype_map,
+            **{pa.large_string(): pa.string(), pa.large_binary(): pa.binary()},
+        }
+
+    def dtype_to_delta_dtype(dtype: pa.DataType) -> pa.DataType:
+        # Handle nested types
+        if isinstance(dtype, (pa.LargeListType, pa.ListType)):
+            return list_to_delta_dtype(dtype)
+        elif isinstance(dtype, pa.StructType):
+            return struct_to_delta_dtype(dtype)
+        elif isinstance(dtype, pa.TimestampType):
+            return pa.timestamp("us")
+        try:
+            return dtype_map[dtype]
+        except KeyError:
+            return dtype
 
-    table = pa.Table.from_pandas(data)
-    schema = table.schema
-    schema_out = []
-    for field in schema:
-        if isinstance(field.type, pa.TimestampType):
-            f = pa.field(
-                name=field.name,
-                type=pa.timestamp("us"),
-                nullable=field.nullable,
-                metadata=field.metadata,
-            )
-            schema_out.append(f)
+    def list_to_delta_dtype(
+        dtype: Union[pa.LargeListType, pa.ListType],
+    ) -> Union[pa.LargeListType, pa.ListType]:
+        nested_dtype = dtype.value_type
+        nested_dtype_cast = dtype_to_delta_dtype(nested_dtype)
+        if large_dtypes:
+            return pa.large_list(nested_dtype_cast)
         else:
-            schema_out.append(field)
-    schema = pa.schema(schema_out, metadata=schema.metadata)
-    table = table.cast(target_schema=schema)
-    return table, schema
+            return pa.list_(nested_dtype_cast)
+
+    def struct_to_delta_dtype(dtype: pa.StructType) -> pa.StructType:
+        fields = [dtype.field(i) for i in range(dtype.num_fields)]
+        fields_cast = [pa.field(f.name, dtype_to_delta_dtype(f.type)) for f in fields]
+        return pa.struct(fields_cast)
+
+    return pa.schema([pa.field(f.name, dtype_to_delta_dtype(f.type)) for f in schema])
+
+
+def convert_pyarrow_recordbatchreader(
+    data: pa.RecordBatchReader, large_dtypes: bool
+) -> pa.RecordBatchReader:
+    """Converts a PyArrow RecordBatchReader to a PyArrow RecordBatchReader with a compatible delta schema"""
+    schema = _convert_pa_schema_to_delta(data.schema, large_dtypes=large_dtypes)
+    data = pa.RecordBatchReader.from_batches(
+        schema,
+        data.read_all().cast(schema).to_batches(),
```

Comment:
It would be nice to achieve this with a generator rather that loading the entire reader into memory.

Reply:
I was looking into it but I don't see a way on how I could change the schema of the generator inside the reader. I tried to go from reader -> scanner but that's a dead end. Reader -> dataset, should theoretically work but `pa.dataset.dataset `mentions RecordBatchReader is an unsupported type even though the docs mentions it is a valid input type.

I am also not sure if materializing is even an issue because later before write or merge we do `valide_batch` which seems to materialize the results.

Reply:
`validate_batch` is run per batch within a reader. `pa.RecordBatchReader.from_batches` will not materialize the batches if you pass a generator, like we do here:

https://github.com/delta-io/delta-rs/blob/9b938308490a60808443ee36b4f7b2c3055a6c15/python/deltalake/writer.py#L341-L343

So this should work:

```python
def cast_to_schema(reader):
    for batch in reader:
        yield pa.Table.from_batches([batch]).cast(schema).to_batches()[0]
data = pa.RecordBatchReader.from_batches(schema, cast_to_schema(schema))
```

Reply:
I see, I thought going through the reader is materializing the results. Ben then I should be able to do the same with the dataset once I convert it to a reader through the scanner.

Reply:
@wjones127 should be good now